### ISSUE TITLE
Lack of s390x arch support

### ISSUE THREAD

**Issue by lcavassa** [canonical/microk8s#1611]
[inspection-report-20200930_075541.tar.gz](https://github.com/ubuntu/microk8s/files/5305235/inspection-report-20200930_075541.tar.gz)
Please run `microk8s inspect` and attach the generated tarball to this issue.

We appreciate your feedback. Thank you for using microk8s.

**Comment by e-desouza** [comment]
Encountered the same issue:

```sudo snap install --classic microk8s
error: snap "microk8s" is not available on stable for this architecture (s390x) but exists on other
       architectures (amd64, arm64, ppc64el).
```

**Comment by ktsakalozos** [comment]
We have started working on this. Should have it ready soon.

**Comment by kycfeel** [comment]
I'm also looking forward!!! Can't wait to apply the microk8s on my System Z.

**Comment by ktsakalozos** [comment]
Could you try the `latest/edge` channel and let me know of problems you may find? I know many of the addons are not available.

```
sudo snap install microk8s --classic --channel=latest/edge
```

Thanks

**Comment by e-desouza** [comment]
I tried the above & followed the other recommendations [here](https://microk8s.io/docs).

`microk8s kubectl create deployment nginx --image=nginx` was stuck at container creating though.. Do I need to label the node as master/worker? I also tried to enable dns first, but it still stuck at ContainerCreating.

**Comment by ktsakalozos** [comment]
@e-desouza this looks as if containerd did not start properly. Can you please attach a `microk8s inspect` tarball?

**Comment by kycfeel** [comment]
@ktsakalozos Hey. I just built a cluster with 2 System Z VMs by using the `latest/edge` version of microk8s as you said. The cluster is healthy now, but there was an issue while I was bootstrapping it.

All pods except for the `calico-node` were returning the error below and stuck in `ContainerCreating` status. 

```
  Warning  FailedCreatePodSandBox  2m10s (x167 over 40m)  kubelet            (combined from similar events): Failed to create pod sandbox: rpc error: code = Unknown desc = failed to setup network for sandbox "577659974a2f42edad7bf0637805b72b69f88dfe67e7934ee9223d86fab62b4f": stat /var/lib/calico/nodename: no such file or directory: check that the calico/node container is running and has mounted /var/lib/calico/
 ```

They were complaining that the `/var/lib/calico/` is missing on host machines. I dived in for a bit more and found out that the `calico-node` pods were only mounting into `/var/snap/microk8s/current/var/lib/` for the `calico` directory on hosts. 

I manually created symlinks on each machine to bind `/var/snap/microk8s/current/var/lib/calico` to `/var/lib/calico`, and all other pods started to get created properly right away.

This is my first time trying `microk8s` so not sure how it behaves on normal amd64 machines, but pretty sure it's a bug or something 🤔.

Other than that, the system health is good so far.

**Comment by ktsakalozos** [comment]
That is strange. If you look at [1] and [2] you will see that the calico node is mounting the host path `/var/snap/microk8s/current/var/lib/calico` under `/var/lib/calico`. 

> All pods except for the `calico-node` were returning the error

Can you give me an example of a pod that is failing? What workload are you trying?

[1] https://github.com/ubuntu/microk8s/blob/master/upgrade-scripts/000-switch-to-calico/resources/calico.s390x.yaml#L3976
[2] https://github.com/ubuntu/microk8s/blob/master/upgrade-scripts/000-switch-to-calico/resources/calico.s390x.yaml#L3989

**Comment by kycfeel** [comment]
@ktsakalozos Hi. The "all pods" that I mentioned include something like `kubernetes-dashboard` or even `coredns`.

Not sure why that hostpath mount issue happened. The `/var/snap/microk8s/current/var/lib/calico` was not under the `/var/lib/calico` on my system tho. Those two directories were fully separated with no symlinks. Is there a chance it's a bug with the latest/edge?

**Comment by ktsakalozos** [comment]
I cannot confirm your findings @kycfeel, latest/edge deployed without problems in a test I just did.

**Comment by kycfeel** [comment]
Hmm... I'm with 3 Ubuntu 18.04 LPARs (VMs) on IBM Rockhopper ii mainframe. And this issue happened on every LPARs I got. Did you also tried to replicate it with s390x machine?

**Comment by ktsakalozos** [comment]
@lcavassa could you speak on how the machine we are testing on is setup?

@kycfeel would it be possible to attach a `microk8s.inspect` tarball.

**Comment by lorenzo-cavassa** [comment]
Is it clustered? I never tried yet to cluster it. 1 LPAR only.

**Comment by frank-heimes** [comment]
We use a single LPAR on a z13 and run 20.04.2 LTS (with latest patches).
System has 32GB RAM, about 100GB disk space and two qeth based network devices.

**Comment by kycfeel** [comment]
Hey guys. The #2353 PR helped me to solve the pod `ContainerCreating` issue. Thank you all.

Not sure this is relevant, but I'm also having a DNS resolving issue between nodes.

The nodes can't communicate with each other with their host name. 

The cluster returns "Temporary failure in name resolution" error If I run commands which requires an inter-cluster communication, like `kubectl exec`.

Here's an example:

```
kycfeel-Macbook:deploy kycfeel$ kubectl exec deploy-heketi-654dbb64df-whbj8 -- curl http://localhost
Error from server: error dialing backend: dial tcp: lookup mycluster-node3: Temporary failure in name resolution
```

Do you guys also see this kinda DNS issue too?

**Comment by kycfeel** [comment]
@ktsakalozos Hey there. I have a quick question. Is there a way to install the v1.21 Kubernetes on `s390x`? 

Based on what I'm seeing, the `edge` channel only got the `v1.22` of Kubernetes right now. 

But since there are some incompatibility issues between our applications and the v1.22 Kubernetes, I want to downgrade or reinstall the microk8s to v1.21 or lower. 

I already tried something like `snap install microk8s --classic --channel=1.21/stable`, but there was no `1.21` channel available for s390x machines for now. 

Can you tell me how can I install the v1.21 microk8s for the IBM mainframe?

Thanks as always :)

**Comment by e-desouza** [comment]
Hi @ktsakalozos, will Canonical provide support for this for enterprise clients who pay for Ubuntu S&S?

**Comment by ktsakalozos** [comment]
Hi @e-desouza, yes Canonical will provide support for clients, enterprise or not.

Let me give an update on where we are and what the next steps on s390x are.

You can now get MicroK8s on s390x with 
```
sudo snap install mcirok8s --classic --channel=1.22/edge
```
S390x builds are on the edge channels, latest/edge and 1.22/edge. These channels follow the work we are doing on all architectures. Given we see traction and interest for the s390x builds we will be releasing revisions to the stable channels. Therefore, if you want us to keep working on this architecture please make some noise. Thanks.

**Comment by frank-heimes** [comment]
Well, I think this can even be closed.
iirc microk8s for s390x is available since ~ v1.23.
